File: OpenStreetMap/Shared/UndoManager.m
//
//  UndoManager.m
//  Spider
//
//  Created by Bryce Cogswell on 8/16/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

#import "DLog.h"
#import "UndoManager.h"

@class OsmBaseObject;

static void RunLoopObserverCallBack(CFRunLoopObserverRef observer,CFRunLoopActivity activity,void *info)
{
	if ( activity & kCFRunLoopAfterWaiting ) {
		UndoManager * undoManager = (__bridge id)info;
		++undoManager.runLoopCounter;
	}
}


#pragma mark UndoAction

@interface UndoAction : NSObject <NSCoding>
@property (readonly,nonatomic)	NSString	*	selector;
@property (readonly,nonatomic)	id				target;
@property (readonly,nonatomic)	NSArray		*	objects;
@property (assign,nonatomic)	NSInteger		group;
@end


@implementation UndoAction

-(instancetype)initWithTarget:(id)target selector:(SEL)selector objects:(NSArray *)objects
{
	DbgAssert( target && selector );
	self = [super init];
	if ( self ) {
		_target = target;
		_selector = NSStringFromSelector(selector);
		_objects = objects;
	}
	return self;
}


- (void)encodeWithCoder:(NSCoder *)coder
{
	DbgAssert(_target);
	[coder encodeObject:_target		forKey:@"target"];
	[coder encodeObject:_selector	forKey:@"selector"];
	[coder encodeObject:_objects	forKey:@"objects"];
	[coder encodeInteger:_group		forKey:@"group"];
}

- (id)initWithCoder:(NSCoder *)coder
{
	self = [super init];
	_target		= [coder decodeObjectForKey:@"target"];
	_selector	= [coder decodeObjectForKey:@"selector"];
	_objects	= [coder decodeObjectForKey:@"objects"];
	_group		= [coder decodeIntegerForKey:@"group"];
	DbgAssert(_target);
    return self;
}

-(NSString *)description
{
	return [NSString stringWithFormat:@"%@ %@ %@ (%@) %ld", [super description], _target, _selector, _objects, (long)_group];
}

-(void)performAction
{
	if ( _selector ) {
		// method call
		SEL selector = NSSelectorFromString(_selector);
		assert(_target);
		assert(_objects);
		NSMethodSignature * sig = [_target methodSignatureForSelector:selector];
		assert( sig );
		assert( _objects.count+2 == [sig numberOfArguments] );
		NSInvocation * invocation = [NSInvocation invocationWithMethodSignature:sig];
		invocation.selector = selector;
		invocation.target = _target;
		for ( NSInteger index = 0; index < _objects.count; ++index ) {
			id obj = _objects[index];

			const char * type = [sig getArgumentTypeAtIndex:2+index];
			switch ( *type ) {
				case 'c':
					{
						char c = [obj charValue];
						[invocation setArgument:(void *)&c atIndex:2+index];
					}
					break;
				case 'd':
					{
						double d = [obj doubleValue];
						[invocation setArgument:(void *)&d atIndex:2+index];
					}
					break;
				case 'i':
					{
						int i = (int)[obj integerValue];
						[invocation setArgument:(void *)&i atIndex:2+index];
					}
					break;
				case 'q':
					{
						long long l = [obj longLongValue];
						[invocation setArgument:(void *)&l atIndex:2+index];
					}
					break;
				case 'B':
					{
						BOOL b = [obj boolValue];
						[invocation setArgument:(void *)&b atIndex:2+index];
					}
					break;
				case '@':
					if ( obj == [NSNull null] )
						obj = nil;
					[invocation setArgument:(void *)&obj atIndex:2+index];
					break;
				default:
					assert(NO);
			}
		}
		[invocation invoke];
		return;
	}
	
	// unknown action type
	assert(NO);
}

@end


#pragma mark UndoManager

@implementation UndoManager

@synthesize isUndoing = _isUndoing;
@synthesize isRedoing = _isRedoing;

-(id)init
{
	self = [super init];
	if ( self ) {
		_undoStack = [NSMutableArray array];
		_redoStack = [NSMutableArray array];

		_groupingStack = [NSMutableArray new];
		
		CFRunLoopObserverContext context = { 0 };
		context.info = (__bridge void *)self;
		_runLoopObserver = CFRunLoopObserverCreate( kCFAllocatorDefault, kCFRunLoopAfterWaiting,
																YES, 0, RunLoopObserverCallBack, &context );
		CFRunLoopAddObserver(CFRunLoopGetMain(), _runLoopObserver, kCFRunLoopCommonModes);
//		DLog(@"add observer %@",_runLoopObserver);
	}
	return self;
}

-(void)dealloc
{
	if ( _runLoopObserver ) {
		CFRunLoopRemoveObserver( CFRunLoopGetMain(), _runLoopObserver, kCFRunLoopCommonModes );
//		DLog(@"remove observer %@",_runLoopObserver);
	}
}

-(NSString *)description
{
	NSMutableString * text = [NSMutableString new];
	[text appendFormat:@"undo stack:\n"];
	[text appendFormat:@"   run loop = %ld\n", (long)_runLoopCounter];
	[text appendFormat:@"   group = %@\n", _groupingStack.count ? _groupingStack.lastObject : @"none"];
	for ( UndoAction * action in _undoStack ) {
		[text appendFormat:@"   %ld: %@\n", (long)action.group, action.selector];
	}
	return text;
}

-(BOOL) canUndo
{
	return _undoStack.count > 0;
}
-(BOOL) canRedo
{
	return _redoStack.count > 0;
}

-(NSInteger)count
{
	return _undoStack.count + _redoStack.count;
}

-(void)postChangeNotification
{
	NSNotification * notification = [NSNotification notificationWithName:UndoManagerDidChangeNotification object:self];
	[[NSNotificationQueue defaultQueue] enqueueNotification:notification postingStyle:NSPostWhenIdle coalesceMask:NSNotificationCoalescingOnName|NSNotificationCoalescingOnSender forModes:nil];
}

-(void)removeMostRecentRedo
{
	assert( _redoStack.count );

	NSInteger group = ((UndoAction *)_redoStack.lastObject).group;
	while ( _redoStack.count && ((UndoAction *)_redoStack.lastObject).group == group ) {
		[_redoStack removeLastObject];
	}

	[self postChangeNotification];
}

- (void) removeAllActions
{
	[self willChangeValueForKey:@"canUndo"];
	[self willChangeValueForKey:@"canRedo"];

	assert(!_isUndoing && !_isRedoing);
	[_undoStack removeAllObjects];
	[_redoStack removeAllObjects];

	[self didChangeValueForKey:@"canUndo"];
	[self didChangeValueForKey:@"canRedo"];

	[self postChangeNotification];
}

-(void)registerUndo:(UndoAction *)action
{
	action.group = _groupingStack.count ? [_groupingStack.lastObject integerValue] : self.runLoopCounter;

	[self willChangeValueForKey:@"canUndo"];
	[self willChangeValueForKey:@"canRedo"];
	
	if ( _isUndoing ) {
		[_redoStack addObject:action];
	} else if ( _isRedoing ) {
		[_undoStack addObject:action];
	} else {
		[_undoStack addObject:action];
		[_redoStack removeAllObjects];
	}

	[self didChangeValueForKey:@"canUndo"];
	[self didChangeValueForKey:@"canRedo"];

	[self postChangeNotification];
}

- (void)registerUndoWithTarget:(id)target selector:(SEL)selector objects:(NSArray *)objects
{
	assert(target);
	DbgAssert( [target respondsToSelector:selector] );

	UndoAction * action = [[UndoAction alloc] initWithTarget:target selector:selector objects:objects];
	[self registerUndo:action];
}

- (void)doComment:(NSDictionary *)comment
{
	[self registerUndoWithTarget:self selector:@selector(doComment:) objects:@[comment]];
	[_commentList addObject:comment];
}

- (void)registerUndoComment:(NSDictionary *)comment
{
	[self registerUndoWithTarget:self selector:@selector(doComment:) objects:@[comment]];
}

+(void)doActionGroupFromStack:(NSMutableArray *)stack
{
	NSInteger currentGroup = -1;
	
	while ( stack.count ) {
		UndoAction * action = stack.lastObject;
		assert(action.group);
		if ( currentGroup < 0 ) {
			currentGroup = action.group;
		} else if ( action.group != currentGroup )
			break;

		[stack removeLastObject];

//		DLog(@"-- Undo action: '%@' %@", action.selector, [action.target description] );
		[action performAction];
	}
}

-(NSDictionary *)undo
{
	_commentList = [NSMutableArray new];

	[self willChangeValueForKey:@"canUndo"];
	[self willChangeValueForKey:@"canRedo"];

	assert(!_isUndoing && !_isRedoing);
	_isUndoing = YES;
	[UndoManager doActionGroupFromStack:_undoStack];
	_isUndoing = NO;

	[self didChangeValueForKey:@"canUndo"];
	[self didChangeValueForKey:@"canRedo"];

	[self postChangeNotification];

	NSDictionary * comment = _commentList.count == 0 ? nil : _commentList.lastObject;
	return comment;
}

-(NSDictionary *)redo
{
	_commentList = [NSMutableArray new];

	[self willChangeValueForKey:@"canUndo"];
	[self willChangeValueForKey:@"canRedo"];

	assert(!_isUndoing && !_isRedoing);
	_isRedoing = YES;
	[UndoManager doActionGroupFromStack:_redoStack];
	_isRedoing = NO;

	[self didChangeValueForKey:@"canUndo"];
	[self didChangeValueForKey:@"canRedo"];

	[self postChangeNotification];
	
	NSDictionary * comment = _commentList.count == 0 ? nil: _commentList[0];
	return comment;
}

-(void)beginUndoGrouping
{
	NSNumber * group = _groupingStack.count ? _groupingStack.lastObject : @(self.runLoopCounter);
	[_groupingStack addObject:group];
}
-(void)endUndoGrouping
{
	[_groupingStack removeLastObject];
}

-(NSSet *)objectRefs
{
	NSMutableSet * refs = [NSMutableSet new];
	for ( UndoAction * action in _undoStack ) {
		[refs addObject:action.target];
		[refs addObjectsFromArray:action.objects];
	}
	for ( UndoAction * action in _redoStack ) {
		[refs addObject:action.target];
		[refs addObjectsFromArray:action.objects];
	}
	return refs;
}


- (void)encodeWithCoder:(NSCoder *)coder
{
	if ( [coder allowsKeyedCoding] ) {
		[coder encodeObject:_undoStack forKey:@"undoStack"];
		[coder encodeObject:_redoStack forKey:@"redoStack"];
		[coder encodeInteger:_runLoopCounter forKey:@"runLoopCounter"];
	} else {
		[coder encodeObject:_undoStack];
		[coder encodeObject:_redoStack];
		[coder encodeBytes:&_runLoopCounter length:sizeof _runLoopCounter];
	}
}
- (id)initWithCoder:(NSCoder *)coder
{
	self = [self init];
	if ( [coder allowsKeyedCoding] ) {
		_undoStack = [coder decodeObjectForKey:@"undoStack"];
		_redoStack = [coder decodeObjectForKey:@"redoStack"];
		_runLoopCounter = [coder decodeIntegerForKey:@"runLoopCounter"];
	} else {
		NSUInteger len;
		_undoStack = [coder decodeObject];
		_redoStack = [coder decodeObject];
		_runLoopCounter = *(NSInteger *)[coder decodeBytesWithReturnedLength:&len];
	}
    return self;
}

@end

File: src/Shared/UndoManager.swift
//  Converted to Swift 5.4 by Swiftify v5.4.27034 - https://swiftify.com/
//
//  UndoManager.swift
//  Go Map!
//
//  Created by Bryce Cogswell on 8/16/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

import Foundation

private var UndoManagerDidChangeNotification = "UndoManagerDidChangeNotification"
typealias UndoManagerChangeCallback = () -> Void

class UndoManager: NSObject, NSCoding {
    var runLoopObserver: CFRunLoopObserver?
    var undoStack: [UndoAction] = []
    var redoStack: [UndoAction] = []
    
    
    var groupingStack: [AnyHashable]? // for explicit grouping
    var commentList: NSMutableArray?
    
    private(set) var isUndoing = false
    private(set) var isRedoing = false
    
    var canUndo: Bool {
        return (undoStack?.count ?? 0) > 0
    }
    
    var canRedo: Bool {
        return (redoStack?.count ?? 0) > 0
    }
    
    var countUndoGroups: Int {
        var count = 0
        var group = -1
        for action in undoStack ?? [] {
            guard let action = action as? UndoAction else {
                continue
            }
            if action.group != group {
                count += 1
                group = action.group
            }
        }
        return count
    }
    var runLoopCounter = 0
    
    override init() {
        super.init()
        undoStack = []
        redoStack = []
        
        groupingStack = []
        
        var context = CFRunLoopObserverContext()
        context.info = self
        runLoopObserver = CFRunLoopObserverCreate(
            kCFAllocatorDefault,
            CFRunLoopActivity.afterWaiting.rawValue,
            true,
            CFIndex(0),
            RunLoopObserverCallBack,
            &context)
        CFRunLoopAddObserver(CFRunLoopGetMain(), runLoopObserver, CFRunLoopMode.commonModes)
        //		DLog(@"add observer %@",_runLoopObserver);
    }
    
    deinit {
        if runLoopObserver != nil {
            CFRunLoopRemoveObserver(CFRunLoopGetMain(), runLoopObserver, CFRunLoopMode.commonModes)
            //		DLog(@"remove observer %@",_runLoopObserver);
        }
    }
    
    override var description: String {
        var text = ""
        text += "undo stack:\n"
        text += String(format: "   run loop = %ld\n", runLoopCounter)
        text += "   group = \(((groupingStack?.count ?? 0) != 0 ? (groupingStack?.last as? String) : "none") ?? "")\n"
        for action in undoStack ?? [] {
            guard let action = action as? UndoAction else {
                continue
            }
            text += String(format: "   %ld: %@\n", action.group, action.selector ?? "")
        }
        return text
    }
    
    func count() -> Int {
        return (undoStack?.count ?? 0) + (redoStack?.count ?? 0)
    }
    
    func postChangeNotification() {
        let notification = Notification(name: NSNotification.Name(UndoManagerDidChangeNotification), object: self)
        NotificationQueue.default.enqueue(notification, postingStyle: .whenIdle, coalesceMask: [.onName, .onSender], forModes: nil)
    }
    
    func removeMostRecentRedo() {
        assert((redoStack?.count ?? 0))
        
        let group = (redoStack?.last as? UndoAction)?.group ?? 0
        while (redoStack?.count ?? 0) != 0 && (redoStack?.last as? UndoAction)?.group == group {
            redoStack?.removeLast()
        }
        
        postChangeNotification()
    }
    
    func removeAllActions() {
        willChangeValue(forKey: "canUndo")
        willChangeValue(forKey: "canRedo")
        
        assert(!isUndoing && !isRedoing)
        undoStack?.removeAll()
        redoStack?.removeAll()
        
        didChangeValue(forKey: "canUndo")
        didChangeValue(forKey: "canRedo")
        
        postChangeNotification()
    }
    
    func registerUndo(_ action: UndoAction?) {
        action?.group = (groupingStack?.count ?? 0) != 0 ? (groupingStack?.last as? NSNumber)?.intValue ?? 0 : runLoopCounter
        
        willChangeValue(forKey: "canUndo")
        willChangeValue(forKey: "canRedo")
        
        if isUndoing {
            if let action = action {
                redoStack?.append(action)
            }
        } else if isRedoing {
            if let action = action {
                undoStack.append(action)
            }
        } else {
            if let action = action {
                undoStack.append(action)
            }
            redoStack?.removeAll()
        }
        
        didChangeValue(forKey: "canUndo")
        didChangeValue(forKey: "canRedo")
        
        postChangeNotification()
    }
    
    func registerUndo(withTarget target: AnyObject?, selector: Selector, objects: NSArray) {
        assert(target != nil)
        DbgAssert(target?.responds(to: selector) ?? false)
        
        let action = UndoAction(target: target, selector: selector, objects: objects)
        registerUndo(action)
    }
    
    @objc func doComment(_ comment: NSDictionary) {
        registerUndo(withTarget: self, selector: #selector(doComment(_:)), objects: [comment])
        commentList?.add(comment)
    }
    
    func registerUndoComment(_ comment: NSDictionary) {
        registerUndo(withTarget: self, selector: #selector(doComment(_:)), objects: [comment])
    }
    
    class func doActionGroup(fromStack stack: inout [UndoAction]) {
        var currentGroup = -1

        while stack.count != 0 {
            let action = stack.last
            assert(action?.group != nil)
            if currentGroup < 0 {
                currentGroup = action?.group ?? 0
            } else if action?.group != currentGroup {
                break
            }
            
            stack.removeLast()
            
            //		DLog(@"-- Undo action: '%@' %@", action.selector, [action.target description] );
            action?.perform()
        }
    }
    
    // returns the oldest comment registered within the undo group
    func undo() -> NSDictionary? {
        commentList = NSMutableArray()

        willChangeValue(forKey: "canUndo")
        willChangeValue(forKey: "canRedo")
        
        assert(!isUndoing && !isRedoing)
        isUndoing = true
        UndoManager.doActionGroup(fromStack: &undoStack)
        isUndoing = false
        
        didChangeValue(forKey: "canUndo")
        didChangeValue(forKey: "canRedo")
        
        postChangeNotification()
        
        let comment = ((commentList?.count ?? 0) == 0 ? nil : commentList?.lastObject) as? NSDictionary
        return comment
    }
    
    func redo() -> NSDictionary? {
        commentList = NSMutableArray()
        
        willChangeValue(forKey: "canUndo")
        willChangeValue(forKey: "canRedo")
        
        assert(!isUndoing && !isRedoing)
        isRedoing = true
        UndoManager.doActionGroup(fromStack: &redoStack)
        isRedoing = false
        
        didChangeValue(forKey: "canUndo")
        didChangeValue(forKey: "canRedo")
        
        postChangeNotification()
        
        let comment = ((commentList?.count ?? 0) == 0 ? nil : commentList?.lastObject) as? NSDictionary
        return comment
    }
    
    func beginUndoGrouping() {
        let group: NSNumber? = ((groupingStack?.count ?? 0) != 0 ? groupingStack?.last : NSNumber(value: runLoopCounter)) as? NSNumber
        if let group = group {
            groupingStack?.append(group)
        }
    }
    
    func endUndoGrouping() {
        groupingStack?.removeLast()
    }
    
    func objectRefs() -> NSSet? {
        var refs = NSMutableSet()
        for action in undoStack {
            refs.add(action.target)
            action.objects
//            refs.addObjects(from: )
        }
        for action in redoStack {
            refs.add(action.target)
            refs.addObjects(from: action.objects)
        }
        return refs
    }
    
    func encode(with coder: NSCoder) {
        if coder.allowsKeyedCoding {
            coder.encode(undoStack, forKey: "undoStack")
            coder.encode(redoStack, forKey: "redoStack")
            coder.encode(runLoopCounter, forKey: "runLoopCounter")
        } else {
            coder.encode(undoStack)
            coder.encode(redoStack)
            coder.encodeBytes(&runLoopCounter, length: MemoryLayout.size(ofValue: runLoopCounter))
        }
    }
    
    required convenience init?(coder: NSCoder) {
        self.init()
        if coder.allowsKeyedCoding {
            undoStack = coder.decodeObject(forKey: "undoStack") as? [UndoAction] ?? []
            redoStack = coder.decodeObject(forKey: "redoStack") as? [UndoAction] ?? []
            runLoopCounter = coder.decodeInteger(forKey: "runLoopCounter")
        } else {
            var len: Int
            undoStack = coder.decodeObject() as? [UndoAction] ?? []
            redoStack = coder.decodeObject() as? [UndoAction] ?? []
            runLoopCounter = coder.decodeBytes(withReturnedLength: &len)
            runLoopCounter = Int(coder.decodeBytes(withReturnedLength: UnsafeMutablePointer(&len)) ?? 0)
        }
    }
}

private func RunLoopObserverCallBack(_ observer: CFRunLoopObserver?, _ activity: CFRunLoopActivity, _ info: UnsafeMutableRawPointer?) {
    if activity.rawValue & CFRunLoopActivity.afterWaiting.rawValue != 0 {
        let undoManager = info as? UndoManager
        undoManager?.runLoopCounter = (undoManager?.runLoopCounter ?? 0) + 1
    }
}

// MARK: UndoAction
class UndoAction: NSObject, NSCoding {
    private(set) var selector: String?
    private(set) var target: Any?
    private(set) var objects: NSArray?
    var group = 0
    
    init(target: Any?, selector: Selector, objects: NSArray?) {
        DbgAssert(target != nil && selector != nil)
        super.init()
        self.target = target
        self.selector = NSStringFromSelector(selector)
        self.objects = objects
    }
    
    func encode(with coder: NSCoder) {
        DbgAssert(target != nil)
        coder.encode(target, forKey: "target")
        coder.encode(selector, forKey: "selector")
        coder.encode(objects, forKey: "objects")
        coder.encode(group, forKey: "group")
    }
    
    required init?(coder: NSCoder) {
        super.init()
        target = coder.decodeObject(forKey: "target")
        selector = coder.decodeObject(forKey: "selector") as? String
        objects = coder.decodeObject(forKey: "objects")
        group = coder.decodeInteger(forKey: "group")
        DbgAssert(target != nil)
    }
    
    override var description: String {
        if let target = target {
            return String(format: "%@ %@ %@ (%@) %ld", super.description, target, selector ?? "", objects ?? [], group)
        }
        return ""
    }
    
    func perform() {
        if self.selector != nil {
            // method call
            let selector = NSSelectorFromString(self.selector ?? "")
            assert(target != nil)
            assert(objects != nil)
            let sig = target?.methodSignature(for: selector)
            assert(sig)
            assert((objects?.count ?? 0) + 2 == sig?.numberOfArguments())
            let invocation = NSInvocation(methodSignature: sig)
            invocation.selector = selector
            invocation.target = target
            for index in 0..<(objects?.count ?? 0) {
                var obj = objects?[index]
                
                let type = sig?.getArgumentType(at: 2 + index) ?? 0
                switch type {
                case "c":
                    var c = (obj as? NSNumber)?.int8Value ?? 0
                    invocation.setArgument(&c, at: 2 + index)
                case "d":
                    var d = (obj as? NSNumber)?.doubleValue ?? 0.0
                    invocation.setArgument(&d, at: 2 + index)
                case "i":
                    var i = obj?.intValue ?? 0
                    invocation.setArgument(&i, at: 2 + index)
                case "q":
                    var l = (obj as? NSNumber)?.int64Value ?? 0
                    invocation.setArgument(&l, at: 2 + index)
                case "B":
                    var b = (obj as? NSNumber)?.boolValue ?? false
                    invocation.setArgument(&b, at: 2 + index)
                case "@":
                    if obj == nil {
                        obj = nil
                    }
                    invocation.setArgument(&obj, at: 2 + index)
                default:
                    assert(false)
                }
            }
            invocation.invoke()
            return
        }
        
        // unknown action type
        assert(false)
    }
}

// MARK: UndoManager

func DbgAssert(_ x: Bool) {
    assert(x, "unspecified")
}
